framework/src/org/checkerframework/common/value/Range.java

```diff
@@ -0,0 +1,670 @@
+                return new Range(Long.MIN_VALUE, Long.MIN_VALUE);
+            }
+        }
+
+        long resultFrom;
+        long resultTo;
+        // We needn't worry about the overflow issue starting from here.
+        // To facilitate the calculation of the result range, we categorize all the scenarios into 9
+        // different cases:
+        // 1. this: po, right: nn
+        // 2. this: po, right: np
+        // 3. this: po, right: us
+        // 4. this: ne, right: nn
+        // 5. this: ne, right: np
+        // 6. this: ne, right: us
+        // 7. this: us, right: nn
+        // 8. this: us, right: np
+        // 9. this: us, right: us
+        // (note: po=>positive, ne=>negative, us:=>unknown sign, np=>non-positive, nn=>non-negative)
+        // The categorization corresponds to the following control flow branches.
+        // It's not difficult to verify the calculation of each case.
+        if (from > 0 && right.from >= 0) {
+            resultFrom = from / Math.max(right.to, 1);
+            resultTo = to / Math.max(right.from, 1);
+        } else if (from > 0 && right.to <= 0) {
+            resultFrom = to / Math.min(right.to, -1);
+            resultTo = from / Math.min(right.from, -1);
+        } else if (from > 0) {
+            resultFrom = -to;
+            resultTo = to;
+        } else if (to < 0 && right.from >= 0) {
+            resultFrom = from / Math.max(right.from, 1);
+            resultTo = to / Math.max(right.to, 1);
+        } else if (to < 0 && right.to <= 0) {
+            resultFrom = to / Math.min(right.from, -1);
+            resultTo = from / Math.min(right.to, -1);
+        } else if (to < 0) {
+            resultFrom = from;
+            resultTo = -from;
+        } else if (right.from >= 0) {
+            resultFrom = from / Math.max(right.from, 1);
+            resultTo = to / Math.max(right.from, 1);
+        } else if (right.to <= 0) {
+            resultFrom = to / Math.min(right.to, -1);
+            resultTo = from / Math.min(right.to, -1);
+        } else {
+            resultFrom = Math.min(from, -to);
+            resultTo = Math.max(-from, to);
+        }
+        return new Range(resultFrom, resultTo);
+    }
+
+    /**
+     * Returns a range that includes all possible values of the remainder of dividing an arbitrary
+     * value in this range by an arbitrary value in the specified range.
+     *
+     * @param right the specified range by which this range is divided
+     * @return the range of the remainder of dividing this range by the specified range. Note that
+     *     this range might not be the smallest range that includes all the possible values.
+     */
+    public Range remainder(Range right) {
+        if (this.isNothing() || right.isNothing()) {
+            return NOTHING;
+        }
+
+        if (right.contains(0)) {
+            // should warn division by zero
+            return EVERYTHING;
+        }
+
+        List<Long> possibleValues =
```

Comment:
I could really use some help regarding the estimation of the `Range.remainder` operation -- it is too complicated to be analysed using the same approach (categorise and then compute accordingly) as in `Range.divide`, however, I am not 100% confident with the current implementation either.